### PR TITLE
Sprint1/add input

### PR DESCRIPTION
**Pull request by martingaston**
This PR adds the `PlayerHuman` module which updates the board state based on valid input, with some error checking thrown into the mix. It communicates with both the UI and IO modules. 

Couple of extra notes: 

1) We're using `args.board` to represent the injected `Board` module, but this creates some confusion with capital-B module board and little-B board state - so I think it's worth course correcting the API there 
2) There is currently duplication between the `IO.output` and the `UI.out` functions - in a future PR I'll push everything to the former. I am not 100% convinced on the current setup of `UI` and think it needs some work.

### CODE REVIEW

File: lib/io.ex [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  def get_position(io \\ :stdio) do
+    input = input(io)
+
+    case Integer.parse(input) do
+      {int, _} -> computerise(int)
+      _ -> get_position(io)
+    end
+  end
+
+  defp computerise(position) do
```
Comment (ssunday) [inline lib/io.ex]: I was going to say use a `z` but it's British vs NA lol

File: lib/player_human.ex [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      {:ok, _move} -> args.board.update(board, position, args.mark)
+      {:error, message} -> move(board, args.ui.message(message), args)
+    end
+  end
+
+  def valid_move?(position, _board, _args) when not is_integer(position) do
+    {:error, :nan}
+  end
+
+  def valid_move?(position, _board, _args) when position < 0 or position > 8 do
```
Comment (ssunday) [inline lib/player_human.ex]: consts for 0/8?

File: lib/UI.ex [hunk 3]
```diff
@@ -9,6 +9,15 @@ defmodule UI do
     print_header(io)
   end
 
+  def message(key) do
+    case key do
+      :nan -> "Sorry, that's not a valid number. Please enter a whole number."
```
Comment (ssunday) [inline lib/UI.ex]: nice

File: lib/io.ex [hunk 4]
```diff
@@ -0,0 +1,22 @@
+defmodule TicTacToe.Io do
+  def get_position(io \\ :stdio) do
+    input = input(io)
```
Comment (ssunday) [inline lib/io.ex]: input(io) |> parse |> case

File: lib/player_human.ex [hunk 5]
```diff
@@ -0,0 +1,27 @@
+defmodule PlayerHuman do
+  def move(board, message, args, io \\ :stdio) do
+    args.io.output(io, message)
+    position = args.io.get_position(io)
+
```
Comment (ssunday) [inline lib/player_human.ex]: args.io.get_position(io) |> case

File: lib/io.ex (hunk 4 above)
Comment (martingaston) [reply lib/io.ex]: you can pipe into a case? holy potatoes that's blowing my mind 💥

File: lib/player_human.ex [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      {:error, message} -> move(board, args.ui.message(message), args)
+    end
+  end
+
+  def valid_move?(position, _board, _args) when not is_integer(position) do
+    {:error, :nan}
+  end
+
+  def valid_move?(position, _board, _args)
+      when position < @first_square or position > @last_square do
```
Comment (ssunday) [inline lib/player_human.ex]: could this be extracted out?

**Pull request merged**